templates/base/head.tmpl

```diff
@@ -29,7 +29,7 @@
 	{{template "base/head_style" .}}
 	{{template "custom/header" .}}
 </head>
-<body hx-headers='{"x-csrf-token": "{{.CsrfToken}}"}' hx-swap="outerHTML" hx-ext="morph" hx-push-url="false">
+<body hx-headers='{"x-csrf-token": "{{.CsrfToken}}"}' hx-swap="outerHTML" hx-ext="morph" hx-push-url="false" hx-ext="ws" ws-connect="/-/ws">
```

Comment:
I wonder, how does the websocket client filter that this message is for him? We will be sending many different websocket messages eventually, so we need some way to distinguish in htmx based on message content.

Reply:
Ok. If we could define this filter as a HTML attribute, it would be ideal.

Reply:
`htmx:wsBeforeMessage` could be that https://htmx.org/extensions/web-sockets/

Reply:
Yes, that sounds good. An then define the filter using a custom attribute like `ws-filter`.

Reply:
> Yes, that sounds good. An then define the filter using a custom attribute like ws-filter. It should throw an error if the attribute is missing.

Okay, I think I do not get this one. Maybe a general note to how I implemented this atm:
- a tab connects via websocket to the server `/-/ws`
- the client sends `{ action: "subscrib", data: { url: "http://localhost:3000/anbraten/test/issues/1"}}` to let the server know it is interested in updates regarding this kind of page 
- another client triggers some action (for now deletes a comment)
- the notifier kicks in and the websocket_notifier receives the event
- the websocket_notifier checks for all connected sessions if their url is relevant for the issue comment and checks if the user of that session is allowed to receive that notification data
- the websocket_notifier sends the updated dom element (matched based on #id) via websocket to all previously filtered client sessions

Reply:
One concern I have: Will one page initiate multiple websocket handshakes when multiple `ws-connect` to the same URL are present? I think that would be a waste of sockets because everything could be done over one socket. Ideally the socket would even be initiated inside a `SharedWorker` so multiple tabs share one socket, this is how GitHub does it for example, but I'm fine with one socket per tab.

I have a feeling that the htmx websocket implementation may not be meant to be used the way I imagine (1 socket per tab/origin), but instead it's only good for 1 socket per HTML node. It'll definitely introduce additional latency when every HTML node initiates their own websocket handshake which if I recall is at least 3 RTTs.

Reply:
If it works the way I think it does (1 websocket per element), it'd not make sense to have this filter attribute, and instead we should have a uniqe `ws-connect` URL per element.

Reply:
Its is using one connection per tab already.

The url send via subscribe message makes sure you only get updates relevant for your page. So sth like an updated PR description is not relevant if you are on a users profile for example.

Reply:
I think we can leverage [`createWebSocket`](https://htmx.org/extensions/web-sockets/#configuration):

> createWebSocket - a factory function that can be used to create a custom WebSocket instances. Must be a function, returning WebSocket object

So we initiate a websocket on page load and whenever htmx wants to create a websocket, we give it the already open websocket instead. With this, it should be easy to be moved to a `SharedWorker` later.

Reply:
Implemented the `createWebSocket` function with the same logic as the default from htmx in `htmx.js`. However I don't understand the current benefit of adding this.

Reply:
With only one consumer element there is no benefit, but with multiples, there will be. For it to work, we need to store all open sockets per URL, I will look into adding this.

Reply:
It should work quite well with a single "consumer element" as you can use https://htmx.org/attributes/hx-swap-oob/ to replace any object in the dom. That's what I did with:

```go
// htmxAddElementEnd = "<div hx-swap-oob=\"beforebegin:%s\">%s</div>"
// htmxUpdateElement = "<div hx-swap-oob=\"outerHTML:%s\">%s</div>"
var htmxRemoveElement = "<div hx-swap-oob=\"delete:%s\"></div>"
```

However anything else I should adjust here in this PR?